include/nbl/video/IDescriptorPool.h

```diff
@@ -32,86 +30,186 @@ class NBL_API IDescriptorPool : public core::IReferenceCounted, public IBackendO
+                if (allowsFreeing)
+                {
+                    assert(generalAllocatorReservedSpace);
+                    return generalAllocator.alloc_addr(count, 1u);
+                }
+                else
+                {
+                    return linearAllocator.alloc_addr(count, 1u);
+                }
             }
-            else
+
+            inline void free(const uint32_t allocatedOffset, const uint32_t count)
             {
-                m_linearAllocators[asset::EDT_COUNT] = core::LinearAddressAllocator<uint32_t>(nullptr, 0u, 0u, 1u, m_maxDescriptorCount[asset::EDT_COMBINED_IMAGE_SAMPLER]);
+                assert(generalAllocatorReservedSpace);
+                generalAllocator.free_addr(allocatedOffset, count);
             }
 
-            // Initialize the storages.
-            m_textureStorage = std::make_unique<core::StorageTrivializer<core::smart_refctd_ptr<video::IGPUImageView>>[]>(m_maxDescriptorCount[asset::EDT_COMBINED_IMAGE_SAMPLER]);
-            m_mutableSamplerStorage = std::make_unique<core::StorageTrivializer<core::smart_refctd_ptr<video::IGPUSampler>>[]>(m_maxDescriptorCount[asset::EDT_COMBINED_IMAGE_SAMPLER]);
-            m_storageImageStorage = std::make_unique<core::StorageTrivializer<core::smart_refctd_ptr<IGPUImageView>>[]>(m_maxDescriptorCount[asset::EDT_STORAGE_IMAGE] + m_maxDescriptorCount[asset::EDT_INPUT_ATTACHMENT]);
-            m_UBO_SSBOStorage = std::make_unique<core::StorageTrivializer<core::smart_refctd_ptr<IGPUBuffer>>[]>(m_maxDescriptorCount[asset::EDT_UNIFORM_BUFFER] + m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER] + m_maxDescriptorCount[asset::EDT_UNIFORM_BUFFER_DYNAMIC] + m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER_DYNAMIC]);
-            m_UTB_STBStorage = std::make_unique<core::StorageTrivializer<core::smart_refctd_ptr<IGPUBufferView>>[]>(m_maxDescriptorCount[asset::EDT_UNIFORM_TEXEL_BUFFER] + m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER_DYNAMIC]);
-            m_accelerationStructureStorage = std::make_unique<core::StorageTrivializer<core::smart_refctd_ptr<IGPUAccelerationStructure>>[]>(m_maxDescriptorCount[asset::EDT_ACCELERATION_STRUCTURE]);
-        }
-
-        virtual ~IDescriptorPool() {}
-
-        virtual core::smart_refctd_ptr<IGPUDescriptorSet> createDescriptorSet_impl(core::smart_refctd_ptr<const IGPUDescriptorSetLayout>&& layout) = 0;
-        virtual bool freeDescriptorSets_impl(const uint32_t descriptorSetCount, IGPUDescriptorSet* const* const descriptorSets) = 0;
+            inline void reset(const bool allowsFreeing)
+            {
+                if (!allowsFreeing)
+                    linearAllocator.reset();
+                else
+                    generalAllocator.reset();
+            }
 
-        uint32_t m_maxSets;
+            inline uint32_t getAllocatedDescriptorCount(const bool allowsFreeing) const
```

Comment:
you don't need `allowsFreeing` you can deduct that from `generalAllocatorReservedSpace` not being a nullptr

Reply:
Resolved.